app/assets/javascripts/discourse/app/static/prosemirror/core/inputrules.js

```diff
@@ -0,0 +1,133 @@
+import {
+  InputRule,
+  inputRules,
+  smartQuotes,
+  textblockTypeInputRule,
+  wrappingInputRule,
+} from "prosemirror-inputrules";
+
+export function buildInputRules(extensions, schema, includeDefault = true) {
+  const rules = [];
+
+  if (includeDefault) {
+    rules.push(
+      // TODO(renato) smartQuotes should respect `markdown_typographer_quotation_marks`
+      ...smartQuotes,
+      ...[
+        wrappingInputRule(/^\s*>\s$/, schema.nodes.blockquote),
+
+        orderedListRule(schema.nodes.ordered_list),
+        bulletListRule(schema.nodes.bullet_list),
+
+        textblockTypeInputRule(/^```$/, schema.nodes.code_block),
+        textblockTypeInputRule(/^ {4}$/, schema.nodes.code_block),
+
+        headingRule(schema.nodes.heading, 6),
+
+        markInputRule(/\*\*([^*]+)\*\*$/, schema.marks.strong),
+        markInputRule(/(?<=^|\s)__([^_]+)__$/, schema.marks.strong),
+        markInputRule(/(?:^|(?<!\*))\*([^*]+)\*$/, schema.marks.em),
+        markInputRule(/(?<=^|\s)_([^_]+)_$/, schema.marks.em),
+        markInputRule(/`([^`]+)`$/, schema.marks.code),
+      ]
+    );
+  }
+
+  rules.push(...extractInputRules(extensions, schema));
+
+  return inputRules({ rules });
+}
+
+function extractInputRules(extensions, schema) {
+  return extensions
+    .reduce((acc, extension) => {
+      if (extension.inputRules) {
+        acc.push(extension.inputRules);
+      }
+      return acc;
+    }, [])
+    .flatMap((inputRule) => processInputRule(inputRule, schema));
+}
+
+function processInputRule(inputRule, schema) {
+  if (inputRule instanceof Array) {
+    return inputRule.map((rule) => processInputRule(rule, schema));
+  }
+
+  if (inputRule instanceof Function) {
+    inputRule = inputRule({ schema, markInputRule });
+  }
+
+  if (inputRule instanceof InputRule) {
+    return inputRule;
+  }
```

Comment:
If I'm understanding correctly, this means that a plugin can provide inputRules in many different ways: a single rule, a callback which returns a single input rule, or a (nested) array of items which can be either functions or rules.

That many options is quite hard to describe & document, so I think it would be better to make things more strict. How about always requiring extensions to provide a callback, and then ensuring that the result is always an array of InputRule? That'll mean every extension will have the same layout, and we only need to document/test one approach.

Reply:
Yes, I've tried to document the different ways in the JSDoc, but we can be a bit more strict here and with the keymap

Reply:
After looking more, I agree it can be nicer to allow non-callback and non-array stuff here.